packages/net/http/_test.pony

```diff
@@ -362,3 +367,196 @@ primitive _Test
+        else 
+          h.log("Error in worker.listening")
+          // h.complete(false)
+        end // try
+
+      be dispose() =>
+        try
+          (client as HTTPClient iso).dispose()
+        end
+
+    end // object
+
+    h.dispose_when_done(worker)
+
+    // Start the fake server.
+    h.dispose_when_done(
+      TCPListener(
+        h.env.root as AmbientAuth,
+        _FixedResponseHTTPServerNotify(
+          h, 
+          {(p: String val) =>
+            worker.listening(p)
+          },
+          recover 
+            [ as String val: 
+              "HTTP/1.1 200 OK"
+              "Server: pony_fake_server"
+              "Content-Length: 0"
+              "Status: 200 OK"
+              ""
+            ]
+          end
+        ),
+        "", // all interfaces
+        "0" // random service
+      )
+    )
+
+    // Start a long test for 5 seconds.
+    h.long_test(5_000_000_000)
+
+primitive _FixedResponseHTTPServerNotify
+  fun apply(
+    h': TestHelper, 
+    f: {(String val)} iso, 
+    r: Array[String val] val)
+    : TCPListenNotify iso^
+  =>
+    recover
+      object is TCPListenNotify iso^
+        let h: TestHelper = h'
+        let listen_cb: {(String val)} iso = consume f
+        let response: Array[String val] val = r
+
+        fun ref listening(listen: TCPListener ref) =>
+          try
+            // Get the service as numeric.
+            let name = listen.local_address().name()?
+            listen_cb(name._2)
+          end
+
+        fun ref not_listening(listen: TCPListener ref) =>
+          h.log("Not listening")
+
+        fun ref closed(listen: TCPListener ref) =>
+          h.log("closed")
+
+        fun ref connected(listen: TCPListener ref): TCPConnectionNotify iso^ =>
+          recover 
+            object is TCPConnectionNotify iso^
+            // let response': Array[String val] val = response
+            let reader: Reader iso = Reader
+
+            fun ref received(
+              conn: TCPConnection ref,
+              data: Array[U8] iso,
+              times: USize)
+              : Bool
+            =>
+              // Test if the request was issued completely.
+              reader.append(consume data)
+              while true do
+                let start = 
+                  try 
+                    let l = reader.line()?
+                    l.contains("HTTP/1.1")
+                  else
+                    break
+                  end
+
+                // Write the response.
+                if start then
+                  for r in response.values() do
+                    conn.write(r + "\n")
+                  end
+                end
+              end // while
+              true
+
+            fun ref accepted(conn: TCPConnection ref) =>
+              None
+
+            fun ref connecting(conn: TCPConnection ref, count: U32) =>
+              None
+            
+            fun ref connect_failed(conn: TCPConnection ref) =>
+              None
```

Comment:
i personally like the approach of https://github.com/ponylang/ponyc/blob/master/packages/net/_test.pony#L202 where multiple `TestHelper.expect_action` calls are made for each stage towards completion of the test. At points like this, where the test failed, a `TestHelper.fail_action` call can greatly help pinpointing possible issues.